couchdb/server/stream.py

```diff
@@ -0,0 +1,56 @@
+# -*- coding: utf-8 -*-
+#
+"""Controls all workflow with input/output streams"""
+import logging
+import sys
+
+from couchdb import json, util
+from couchdb.server.exceptions import FatalError
+
+__all__ = ('receive', 'respond')
+
+log = logging.getLogger(__name__)
+
+
+def receive(input=sys.stdin):
+    """Yields json decoded line from input stream.
+
+    :param input: Input stream with `.readline()` support.
+
+    :yields: JSON decoded object.
+    :rtype: list
+    """
+    while True:
+        line = input.readline()
+        if not line:
+            break
+        log.debug('Input:\n%r', line)
+        try:
+            yield json.decode(line)
+        except Exception as err:
+            log.exception('Unable to decode json data:\n%s', line)
+            raise FatalError('json_decode', str(err))
+
+
+def respond(obj, output=sys.stdout):
+    """Writes json encoded object to output stream.
+
+    :param obj: JSON encodable object.
+    :type obj: dict or list
+
+    :param output: Output file-like object.
+    """
+    if obj is None:
+        log.debug('Nothing to respond')
+        return
+    try:
+        obj = json.encode(obj) + '\n'
+    except Exception as err:
+        log.exception('Unable to encode object to json:\n%r', obj)
+        raise FatalError('json_encode', str(err))
+    else:
+        if isinstance(obj, util.utype):
+            obj = obj.encode('utf-8')
+        log.debug('Output:\n%r', obj)
+        output.write(obj)
+        output.flush()
```

Comment:
I apply the following patch, run the ci and [got some error on py27](https://travis-ci.org/iblis17/couchdb-python/jobs/130407089).
```diff
diff --git a/couchdb/server/stream.py b/couchdb/server/stream.py
index 34c69fd..b79bfc7 100644
--- a/couchdb/server/stream.py
+++ b/couchdb/server/stream.py
@@ -49,8 +49,6 @@ def respond(obj, output=sys.stdout):
         log.exception('Unable to encode object to json:\n%r', obj)
         raise FatalError('json_encode', str(err))
     else:
-        if isinstance(obj, util.utype):
-            obj = obj.encode('utf-8')
         log.debug('Output:\n%r', obj)
         output.write(obj)
         output.flush()
diff --git a/couchdb/tests/server/stream.py b/couchdb/tests/server/stream.py
index cc0d321..3ad825e 100644
--- a/couchdb/tests/server/stream.py
+++ b/couchdb/tests/server/stream.py
@@ -28,9 +28,10 @@ class StreamTestCase(unittest.TestCase):

     def test_respond(self):
         """should encode object to json and write it to output stream"""
+        from io import StringIO
         output = StringIO()
         stream.respond(['foo', {'bar': ['baz']}], output)
-        self.assertEqual(output.getvalue(), b'["foo", {"bar": ["baz"]}]\n')
+        self.assertEqual(output.getvalue(), '["foo", {"bar": ["baz"]}]\n')

     def test_fail_on_respond_unserializable_to_json_object(self):
         """should raise FatalError if json encode fails"""
```

But I cannot reproduce some error like this one:
```
======================================================================
ERROR: test_log_none_message (couchdb.tests.server.qs.BaseQueryServerTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/iblis17/couchdb-python/couchdb/tests/server/qs.py", line 245, in test_log_none_message
    server.log(None)
  File "/home/travis/build/iblis17/couchdb-python/couchdb/server/__init__.py", line 254, in log
    self.respond(res)
  File "/home/travis/build/iblis17/couchdb-python/couchdb/server/__init__.py", line 235, in respond
    return self._respond(data)
  File "/home/travis/build/iblis17/couchdb-python/couchdb/server/stream.py", line 53, in respond
    output.write(obj)
TypeError: 'str' does not support the buffer interface
```